src/Http/Http.Abstractions/src/Extensions/IEndpointConventionBuilder.cs

```diff
@@ -16,4 +16,10 @@ public interface IEndpointConventionBuilder
     /// </summary>
     /// <param name="convention">The convention to add to the builder.</param>
     void Add(Action<EndpointBuilder> convention);
+
+    /// <summary>
+    /// Registers the specified convention for execution after the target builder has been built.
+    /// </summary>
+    /// <param name="finalConvention">The final convention to add to the builder.</param>
+    void Finally(Action<EndpointBuilder> finalConvention) => throw new NotImplementedException();
```

Comment:
Are we worried that making the default method throw will break existing scenarios? Basically, anywhere in the framework where we now call `Finally`, is the `IEndpointConventionBuilder` potentially a user defined one? And if so, was the code path accessible prior to 7.0?

Reply:
> Are we worried that making the default method throw will break existing scenarios?

This is a new API so I'm not too anxious about it breaking existing scenarios and the one place where it is used is `WithOpenApi` which was introduced in 7.0.

> Basically, anywhere in the framework where we now call Finally, is the IEndpointConventionBuilder potentially a user defined one?

Yep, I've added implementations in all the `IEndpointConvenntionBuilder` implementations except those that are used for tests:

- `ComponentEndpointConventionBuilder`
- `RouteHandlerBuilder`
- `ConnectionEndpointRouteBuilder`
- `RouteGroupBuilder`
- `HubEndpointConventionBuilder`
- `ControllerActionEndpointConventionBuilder`
- `PageActionEndpointConventionBuilder`

Reply:
Do a grep.app on github.

Reply:
> Do a grep.app on github.

YARP and gRPC also have implementations of IEndpointConventionBuilder, so we'll want to implement `Finally` there too.

However, at least for the time being, there's no risk of breaking existing scenarios in those APIs since they don't use it currently.

Reply:
It's not about gPRC or YARP or anyone elses implementation using `Finally`, it's about **our** code calling `Finally` on them.

Reply:
> It's not about gPRC or YARP or anyone elses implementation using Finally, it's about our code calling Finally on them.

Yeah, the point above about `WithOpenApi` stands. With this PR, it's the only place that might call into their `Finally`. It's not likely that users will be calling `WithOpenApi` on those builders though and, in the event that they are, it's probably a good thing that NIE is thrown since they don't fully support that API (intentionally).

Reply:
Given that `WithOpenApi` already skips over non `RouteEndpointBuilder` endpoints and endpoints without `MethodInfo` metadata, skipping over unsupported endpoints is already a behavior. Should we swallow the `NotImplemented` exception from `Finally` in `WithOpenApi`? Also, I don't like `finalConvention` because it implies there can only be one of them.

```suggestion
    void Finally(Action<EndpointBuilder> finallyConvention) => throw new NotImplementedException();
```

Reply:
> Should we swallow the NotImplemented exception from Finally in WithOpenApi?

I'm not too enthused about treating the exception thrown here as some sort of contract that `WithOpenApi` will check for.

> endpoints without MethodInfo metadata,

It doesn't generate an operation for these types of endpoints but the `Finally` convention will still be registered on them so it's not totally the same.

> Also, I don't like finalConvention because it implies there can only be one of them.

Bleh. This wan't intentional. I used VS Code to do a find/replace of final => finally in the `src/Http` and `src/Mvc` directories but it looks like this one might've been skipped. :/

Reply:
Or maybe just call it convention because that also makes sense 😄. The property should be final conventions and the parameter should be convention. They are both adding conventions that run a different times.

Reply:
Imagine this was an abstract base class, the options are:
- Add a bool SupportsFinally, return false by default, make the new virtual member throw, or have a reasonable implementation (like calling Add?)